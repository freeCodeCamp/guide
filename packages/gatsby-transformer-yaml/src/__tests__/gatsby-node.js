const Promise = require(`bluebird`)
const yaml = require(`js-yaml`)
const _ = require(`lodash`)

const { onCreateNode } = require(`../gatsby-node`)

describe(`Process YAML nodes correctly`, () => {
  const node = {
    id: `whatever`,
    parent: `SOURCE`,
    children: [],
    internal: {
      contentDigest: `whatever`,
      mediaType: `text/yaml`,
    },
    name: `test`,
  }

  // Make some fake functions its expecting.
  const loadNodeContent = node => Promise.resolve(node.content)

  it(`correctly creates nodes from JSON which is an array of objects`, async () => {
    const data = [{ blue: true, funny: `yup` }, { blue: false, funny: `nope` }]
    node.content = yaml.safeDump(data)

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(2)
      expect(createParentChildLink).toHaveBeenCalledTimes(2)
    })
  })

  it(`If the object has an id, it uses that as the id instead of the auto-generated one`, async () => {
    const data = [
      { id: `foo`, blue: true, funny: `yup` },
      { blue: false, funny: `nope` },
    ]
    node.content = yaml.safeDump(data)

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      expect(createNode.mock.calls[0][0].id).toEqual(`foo`)
    })
  })

  it(`the different objects shouldn't get the same ID even if they have the same content`, async () => {
    const data = [
      { id: `foo`, blue: true, funny: `yup` },
      { blue: false, funny: `nope` },
      { blue: false, funny: `nope` },
      { green: false, funny: `nope` },
    ]
    node.content = yaml.safeDump(data)

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const boundActionCreators = { createNode, createParentChildLink }

    await onCreateNode({
      node,
      loadNodeContent,
      boundActionCreators,
    }).then(() => {
      const ids = createNode.mock.calls.map(object => object[0].id)
      // Test that they're unique
      expect(_.uniq(ids).length).toEqual(4)
    })
  })
})
