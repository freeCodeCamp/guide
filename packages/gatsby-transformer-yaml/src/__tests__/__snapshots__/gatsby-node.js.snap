// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process YAML nodes correctly correctly creates nodes from JSON which is an array of objects 1`] = `
Array [
  Array [
    Object {
      "blue": true,
      "children": Array [],
      "funny": "yup",
      "id": "whatever [0] >>> YAML",
      "internal": Object {
        "content": "{\\"blue\\":true,\\"funny\\":\\"yup\\"}",
        "contentDigest": "73901821b17d5aa9dd6026181f73b64c",
        "mediaType": "application/json",
        "type": "TestYaml",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "blue": false,
      "children": Array [],
      "funny": "nope",
      "id": "whatever [1] >>> YAML",
      "internal": Object {
        "content": "{\\"blue\\":false,\\"funny\\":\\"nope\\"}",
        "contentDigest": "f624311d932d73dcd416d2a8bea2b67d",
        "mediaType": "application/json",
        "type": "TestYaml",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process YAML nodes correctly correctly creates nodes from JSON which is an array of objects 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": true,
        "children": Array [],
        "funny": "yup",
        "id": "whatever [0] >>> YAML",
        "internal": Object {
          "content": "{\\"blue\\":true,\\"funny\\":\\"yup\\"}",
          "contentDigest": "73901821b17d5aa9dd6026181f73b64c",
          "mediaType": "application/json",
          "type": "TestYaml",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "- blue: true
  funny: yup
- blue: false
  funny: nope
",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/yaml",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "blue": false,
        "children": Array [],
        "funny": "nope",
        "id": "whatever [1] >>> YAML",
        "internal": Object {
          "content": "{\\"blue\\":false,\\"funny\\":\\"nope\\"}",
          "contentDigest": "f624311d932d73dcd416d2a8bea2b67d",
          "mediaType": "application/json",
          "type": "TestYaml",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "- blue: true
  funny: yup
- blue: false
  funny: nope
",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "text/yaml",
        },
        "name": "test",
        "parent": "SOURCE",
      },
    },
  ],
]
`;
